### PR TITLE
Implement group join API

### PR DESCRIPTION
**Pull request by dbkr**
Based off https://github.com/matrix-org/synapse/pull/3045

### CODE REVIEW

File: synapse/groups/groups_server.py [hunk 1]
```diff
@@ -719,6 +719,51 @@ def accept_invite(self, group_id, requester_user_id, content):
...
+            is_admin=False,
+            is_public=is_public,
+            local_attestation=local_attestation,
+            remote_attestation=remote_attestation,
+        )
+
+        defer.returnValue({
+            "state": "join",
+            "attestation": local_attestation,
+        })
```
Comment (erikjohnston) [inline synapse/groups/groups_server.py]: We should probably factor out the common code with `accept_invite`

File: synapse/groups/groups_server.py [hunk 2]
```diff
@@ -719,6 +719,51 @@ def accept_invite(self, group_id, requester_user_id, content):
...
+
+        This will error if the group requires an invite/knock to join
+        """
+
+        yield self.check_group_is_ours(group_id, requester_user_id, and_exists=True)
+
+        group_info = yield self.store.get_group(
+            group_id,
+        )
+        if not group_info['is_joinable']:
```
Comment (lukebarnard1) [inline synapse/groups/groups_server.py]: This now has to change

File: synapse/storage/group_server.py [hunk 3]
```diff
@@ -42,19 +42,26 @@ def set_group_joinable(self, group_id, is_joinable):
...
+                "name", "short_description", "long_description",
+                "avatar_url", "is_public", "is_joinable",
             ),
             allow_none=True,
-            desc="is_user_in_group",
+            desc="get_group",
         )
 
+        if ret and 'is_joinable' in ret:
+            ret['is_joinable'] = bool(ret['is_joinable'])
```
Comment (lukebarnard1) [inline synapse/storage/group_server.py]: This now has to change

**Comment by lukebarnard1** [comment]
Let's wait until https://github.com/matrix-org/synapse/pull/3045 has landed before making changes to this again.

File: synapse/storage/group_server.py [hunk 4]
```diff
@@ -48,19 +48,23 @@ def set_group_join_policy(self, group_id, join_policy):
...
-                "name", "short_description", "long_description", "avatar_url", "is_public"
+                "name", "short_description", "long_description",
+                "avatar_url", "is_public", "join_policy",
             ),
             allow_none=True,
-            desc="is_user_in_group",
+            desc="get_group",
         )
 
+        defer.returnValue(ret)
```
[the remark this answers is not in the document]
Comment (lukebarnard1) [reply synapse/storage/group_server.py]: Ah. This is redundant now that we don't do anything with the result before returning.

File: synapse/groups/groups_server.py [hunk 5]
```diff
@@ -678,30 +678,21 @@ def invite_to_group(self, group_id, user_id, requester_user_id, content):
             raise SynapseError(502, "Unknown state returned by HS")
 
     @defer.inlineCallbacks
-    def accept_invite(self, group_id, requester_user_id, content):
-        """User tries to accept an invite to the group.
+    def add_user(self, group_id, user_id, content):
```
Comment (erikjohnston) [inline synapse/groups/groups_server.py]: This should probably be an internal function `_add_user`

File: synapse/storage/group_server.py [hunk 6]
```diff
@@ -55,10 +55,11 @@ def get_group(self, group_id):
                 "group_id": group_id,
             },
             retcols=(
-                "name", "short_description", "long_description", "avatar_url", "is_public"
+                "name", "short_description", "long_description",
+                "avatar_url", "is_public", "join_policy",
```
Comment (erikjohnston) [inline synapse/storage/group_server.py]: Having looked at the sytest I now realise that this gets passed straight through to the client by `/profile`. This API doesn't make much sense as the `join_policy` isn't a string, but a dict.

As you can see the way we do this with visibility is to use a separate flag: we don't specify the precise visibility settings, but simply whether the group is fully public or not. Either we should have an `is_openly_joinable` flag or return the full `join_policy` dict. Personally I'd probably match the existing stuff and return `is_openly_joinable` (or some such).

Comment (lukebarnard1) [reply synapse/storage/group_server.py]: Ok, having discussed this IRL, we're going to go for `is_openly_joinable` as we want to maintain extensibility of the group server concept.

File: synapse/groups/groups_server.py [hunk 7]
```diff
@@ -407,6 +407,11 @@ def get_group_profile(self, group_id, requester_user_id):
         group_description = yield self.store.get_group(group_id)
 
         if group_description:
+            join_policy = group_description['join_policy']
+            del group_description['join_policy']
+
+            group_description['is_openly_joinable'] = join_policy == "open"
```
Comment (erikjohnston) [inline synapse/groups/groups_server.py]: If we're doing this then lets be explicit about what we're returning, e.g.:

```json
group_description = {
    "name": group_description['name'],
    ...
}
```

Mainly to stop us from running into the same problem when adding columns to the DB in future

Comment (lukebarnard1) [reply synapse/groups/groups_server.py]: yep, wfm

File: synapse/groups/groups_server.py [hunk 8]
```diff
@@ -404,9 +404,16 @@ def get_group_profile(self, group_id, requester_user_id):
...
-        group_description = yield self.store.get_group(group_id)
+        group = yield self.store.get_group(group_id)
+
+        if group:
+            cols = [
+                "name", "short_description", "long_description",
+                "avatar_url", "is_public",
+            ]
+            group_description = { key: group[key] for key in cols }
+            group_description["is_openly_joinable"] = join_policy == "open"
```
Comment (krombel) [inline synapse/groups/groups_server.py]: @lukebarnard1 you missed `join_policy = group['join_policy']` or similar

Comment (lukebarnard1) [reply synapse/groups/groups_server.py]: @krombel thank you - I noticed my sytests were failing.

**Pull request merged**